File: sdk/typescript/src/timestamp.ts
import { TimestampExpiredError, TimestampFutureError, TimestampMissingError } from "./errors.js";

export const DEFAULT_MAX_TIMESTAMP_AGE_MS = 5 * 60 * 1000;
export const DEFAULT_CLOCK_SKEW_MS = 30 * 1000;

export function nowMs(): number {
  return Date.now();
}

export function dateFromMs(timestampMs: number): Date {
  return new Date(timestampMs);
}

export function rfc3339SecondsFromMs(timestampMs: number): string {
  const date = dateFromMs(timestampMs);
  if (Number.isNaN(date.getTime())) {
    throw new TimestampMissingError("timestamp is invalid");
  }
  return date.toISOString().replace(/\.\d{3}Z$/u, "Z");
}

export interface TimestampValidatorOptions {
  readonly maxAgeMs?: number;
  readonly clockSkewMs?: number;
}

export class TimestampValidator {
  readonly maxAgeMs: number;
  readonly clockSkewMs: number;

  constructor(options: TimestampValidatorOptions = {}) {
    this.maxAgeMs = options.maxAgeMs ?? DEFAULT_MAX_TIMESTAMP_AGE_MS;
    this.clockSkewMs = options.clockSkewMs ?? DEFAULT_CLOCK_SKEW_MS;
    if (this.maxAgeMs <= 0) {
      throw new TimestampMissingError("maxAgeMs must be positive");
    }
    if (this.clockSkewMs < 0) {
      throw new TimestampMissingError("clockSkewMs cannot be negative");
    }
  }

  validate(timestampMs: number, now: Date = new Date()): void {
    if (timestampMs === 0) {
      throw new TimestampMissingError("timestamp is missing or zero");
    }
    const messageTime = dateFromMs(timestampMs);
    if (Number.isNaN(messageTime.getTime())) {
      throw new TimestampMissingError("timestamp is invalid");
    }
    if (messageTime.getTime() < now.getTime() - this.maxAgeMs) {
      throw new TimestampExpiredError("timestamp expired");
    }
    if (messageTime.getTime() > now.getTime() + this.clockSkewMs) {
      throw new TimestampFutureError("timestamp is too far in the future");
    }
  }

  isExpired(timestampMs: number, now: Date = new Date()): boolean {
    try {
      this.validate(timestampMs, now);
    } catch (error) {
      if (error instanceof TimestampExpiredError) {
        return true;
      }
      throw error;
    }
    return false;
  }
}

File: sdk/python/src/axcp/timestamp.py
"""Timestamp helpers for AXCP replay protection."""

from __future__ import annotations

import dataclasses
from datetime import datetime, timedelta, timezone

from axcp.errors import TimestampExpiredError, TimestampFutureError, TimestampMissingError

DEFAULT_MAX_TIMESTAMP_AGE = timedelta(minutes=5)
DEFAULT_CLOCK_SKEW = timedelta(seconds=30)


def now_ms() -> int:
    return int(datetime.now(tz=timezone.utc).timestamp() * 1000)


def datetime_from_ms(timestamp_ms: int) -> datetime:
    return datetime.fromtimestamp(timestamp_ms / 1000, tz=timezone.utc)


def rfc3339_seconds_from_ms(timestamp_ms: int) -> str:
    return datetime_from_ms(timestamp_ms).strftime("%Y-%m-%dT%H:%M:%SZ")


@dataclasses.dataclass
class TimestampValidator:
    max_age: timedelta = DEFAULT_MAX_TIMESTAMP_AGE
    clock_skew: timedelta = DEFAULT_CLOCK_SKEW

    def validate(self, timestamp_ms: int, now: datetime | None = None) -> None:
        if timestamp_ms == 0:
            raise TimestampMissingError("timestamp is missing or zero")

        current = now or datetime.now(tz=timezone.utc)
        msg_time = datetime_from_ms(timestamp_ms)
        if msg_time < current - self.max_age:
            raise TimestampExpiredError("timestamp expired")
        if msg_time > current + self.clock_skew:
            raise TimestampFutureError("timestamp is too far in the future")

    def is_expired(self, timestamp_ms: int, now: datetime | None = None) -> bool:
        try:
            self.validate(timestamp_ms, now=now)
        except TimestampExpiredError:
            return True
        return False
